File: protocol/src/test/javascript/zfoojs/ProtocolManager.js
import EmptyObject from './packet/EmptyObject.js';
import ComplexObject from './packet/ComplexObject.js';
import NormalObject from './packet/NormalObject.js';
import ObjectA from './packet/ObjectA.js';
import ObjectB from './packet/ObjectB.js';
import SimpleObject from './packet/SimpleObject.js';

const protocols = new Map();

const ProtocolManager = {};

// initProtocol
protocols.set(0, EmptyObject);
protocols.set(100, ComplexObject);
protocols.set(101, NormalObject);
protocols.set(102, ObjectA);
protocols.set(103, ObjectB);
protocols.set(104, SimpleObject);

ProtocolManager.getProtocol = function getProtocol(protocolId) {
    const protocol = protocols.get(protocolId);
    if (protocol === null) {
        throw new Error('[protocolId:' + protocolId + ']协议不存在');
    }
    return protocol;
};

ProtocolManager.write = function write(buffer, packet) {
    const protocolId = packet.protocolId();
    buffer.writeShort(protocolId);
    const protocol = ProtocolManager.getProtocol(protocolId);
    protocol.write(buffer, packet);
};

ProtocolManager.read = function read(buffer) {
    const protocolId = buffer.readShort();
    const protocol = ProtocolManager.getProtocol(protocolId);
    const packet = protocol.read(buffer);
    return packet;
};

export default ProtocolManager;

File: protocol/src/test/lua/zfoolua/ProtocolManager.lua
local protocols = {}

local ProtocolManager = {}

-- table扩展方法，map的大小
function table.mapSize(map)
    local size = 0
    for _,_ in pairs(map) do
        size = size + 1
    end
    return size
end

function ProtocolManager.getProtocol(protocolId)
    local protocol = protocols[protocolId]
    if protocol == nil then
        error("[protocolId:" + protocolId + "] not exist")
    end
    return protocol
end

function ProtocolManager.write(buffer, packet)
    local protocolId = packet:protocolId()
    -- 写入协议号
    buffer:writeShort(protocolId)
    -- 写入包体
    ProtocolManager.getProtocol(protocolId):write(buffer, packet)
end

function ProtocolManager.read(buffer)
    local protocolId = buffer:readShort()
    return ProtocolManager.getProtocol(protocolId):read(buffer)
end

function ProtocolManager.initProtocol()
    local EmptyObject = require("zfoolua.packet.EmptyObject")
    local ComplexObject = require("zfoolua.packet.ComplexObject")
    local NormalObject = require("zfoolua.packet.NormalObject")
    local ObjectA = require("zfoolua.packet.ObjectA")
    local ObjectB = require("zfoolua.packet.ObjectB")
    local SimpleObject = require("zfoolua.packet.SimpleObject")
    protocols[0] = EmptyObject
    protocols[100] = ComplexObject
    protocols[101] = NormalObject
    protocols[102] = ObjectA
    protocols[103] = ObjectB
    protocols[104] = SimpleObject
end

return ProtocolManager